pkg/controllers/node/node_controller.go

```diff
@@ -38,46 +42,162 @@ const (
 	RateLimitCalicoList   = "calico-list"
 	RateLimitK8s          = "k8s"
 	RateLimitCalicoDelete = "calico-delete"
+	nodeLabelAnnotation   = "projectcalico.org/kube-labels"
+)
+
+var (
+	maxAttempts    = 5
+	retrySleepTime = 100 * time.Millisecond
 )
 
 // NodeController implements the Controller interface.  It is responsible for monitoring
 // kubernetes nodes and responding to delete events by removing them from the Calico datastore.
 type NodeController struct {
 	ctx          context.Context
 	informer     cache.Controller
+	indexer      cache.Indexer
 	calicoClient client.Interface
 	k8sClientset *kubernetes.Clientset
 	rl           workqueue.RateLimiter
 	schedule     chan interface{}
+	nodemapper   map[string]string
+	nodemapLock  sync.Mutex
+	syncer       bapi.Syncer
 }
 
 // NewNodeController Constructor for NodeController
-func NewNodeController(ctx context.Context, k8sClientset *kubernetes.Clientset, calicoClient client.Interface) controller.Controller {
+func NewNodeController(ctx context.Context, k8sClientset *kubernetes.Clientset, calicoClient client.Interface, cfg *config.Config) controller.Controller {
+	nc := &NodeController{
+		ctx:          ctx,
+		calicoClient: calicoClient,
+		k8sClientset: k8sClientset,
+		rl:           workqueue.DefaultControllerRateLimiter(),
+		nodemapper:   map[string]string{},
+	}
 	// channel used to kick the controller into scheduling a sync. It has length
 	// 1 so that we coalesce multiple kicks while a sync is happening down to
 	// just one additional sync.
-	schedule := make(chan interface{}, 1)
+	nc.schedule = make(chan interface{}, 1)
 
 	// Create a Node watcher.
 	listWatcher := cache.NewListWatchFromClient(k8sClientset.CoreV1().RESTClient(), "nodes", "", fields.Everything())
 
-	// Informer handles managing the watch and signals us when nodes are deleted.
-	_, informer := cache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, cache.ResourceEventHandlerFuncs{
+	// Setup event handlers
+	handlers := cache.ResourceEventHandlerFuncs{
 		DeleteFunc: func(obj interface{}) {
 			// Just kick controller to wake up and perform a sync. No need to bother what node it was
 			// as we sync everything.
-			kick(schedule)
-		},
-	}, cache.Indexers{})
+			kick(nc.schedule)
+		}}
 
-	return &NodeController{
-		ctx:          ctx,
-		informer:     informer,
-		calicoClient: calicoClient,
-		k8sClientset: k8sClientset,
-		rl:           workqueue.DefaultControllerRateLimiter(),
-		schedule:     schedule,
+	if cfg.SyncNodeLabels {
+		// Start the syncer.
+		nc.initSyncer()
+		nc.syncer.Start()
```

Comment:
I think we end up with a weird race condition where the syncer starts and tries to process updates for the Calico Node objects where it attempts to access the indexer which hasn't been initialized yet. Would switching around the order of initialization fix this?

Reply:
It's intended that it doesn't matter which one occurs first - it should succeed when the other sends an update.